Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/LibraryBuilder.java

```diff
@@ -2935,6 +2985,58 @@ private DataType getExpressionDefResultType(ExpressionDef expressionDef) {
         throw new IllegalArgumentException(String.format("Invalid context reference from %s context to %s context.", currentExpressionContext(), expressionDef.getContext()));
     }
 
+    /**
+     * Add an identifier to the deque to indicate that we are considering it for consideration for identifier hiding and
+     * adding a compiler warning if this is the case.
+     * <p/>
+     * For example, if an alias within an expression body has the same name as a parameter, execution would have
+     * added the parameter identifier and the next execution would consider an alias with the same name, thus resulting
+     * in a warning.
+     * <p/>
+     * Exact case matching as well as case-insensitive matching are considered.  If known, the type of the structure
+     * in question will be considered in crafting the warning message, as per the {@link Element} parameter.
+     *
+     * @param identifier The identifier belonging to the parameter, expression, function, alias, etc, to be evaluated.
+     * @param onlyOnce   Special case to deal with overloaded functions, which are out scope for hiding.
+     * @param element    The construct element, for {@link ExpressionRef}.
+     */
+    void pushIdentifierForHiding(String identifier, boolean onlyOnce, Element element) {
+        final MatchType matchType = identifiersToCheckForHiding.stream()
```

Comment:
I am a little concerned about performance here. For library-level identifier resolution a hash is used currently. If I'm reading this correctly, every identifier declaration anywhere will now incur a worst-case performance of O(3n) in the number of identifiers in the library. Since that hit is on every identifier declaration, that will add up quickly. For most libraries, it probably doesn't matter, but I'm concerned about libraries with large numbers of declarations, and/or lots of content. It would be pretty straightforward to recast this as a hash, even for the case-insensitive comparison, right? Or am I being overly cautious here?

Reply:
I think it's probably overly cautious, myself.

The reason for that is that the alternative still requires a Deque or LinkedList or some other such construct to represent identifier scoping (for parameters, aliases, and lets). A hash alone won't cut it because we'd need to pop the identifiers off the stack once they fall out of scope. Best case is a linked list of hashes, which all need to be checked.

That would actually be the same construct we use at runtime in the cql engine to represent the stack frame, and if it's "fast enough" at runtime it ought to be "fast enough" at compile time, IMO.

I'm also not sure where O(3n) comes from. Could you elaborate on that? I think it's actually worse than that, O(N^2) (check each identifier against every other identifier).

We don't have any benchmarking for the compiler in the project, only the engine. It'd be interesting to add a few tests and see where our hotspots are. How about we add a story to include some benchmarks, and we address whatever hotspots there may be? If it's this particular item we change it, if it turns out to be something else we can change that.